File: app/src/main/java/com/robifr/ledger/ui/selectcustomer/recycler/SelectCustomerAdapter.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.selectcustomer.recycler;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.robifr.ledger.data.model.CustomerModel;
import com.robifr.ledger.databinding.CustomerCardWideBinding;
import com.robifr.ledger.databinding.ListableListSelectedItemBinding;
import com.robifr.ledger.ui.RecyclerViewHolder;
import com.robifr.ledger.ui.selectcustomer.SelectCustomerFragment;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;

public class SelectCustomerAdapter extends RecyclerView.Adapter<RecyclerViewHolder> {
  private enum ViewType {
    HEADER(0),
    LIST(1);

    private final int _value;

    private ViewType(int value) {
      this._value = value;
    }

    public int value() {
      return this._value;
    }
  }

  @NonNull private final SelectCustomerFragment _fragment;

  public SelectCustomerAdapter(@NonNull SelectCustomerFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
  }

  @Override
  @NonNull
  public RecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    Objects.requireNonNull(parent);

    final ViewType type =
        Arrays.stream(ViewType.values())
            .filter(e -> e.value() == viewType)
            .findFirst()
            .orElse(ViewType.LIST);
    final LayoutInflater inflater = this._fragment.getLayoutInflater();

    return switch (type) {
      case HEADER ->
          new SelectCustomerHeaderHolder(
              this._fragment, ListableListSelectedItemBinding.inflate(inflater, parent, false));

        // Defaults to `ViewType#LIST`.
      default ->
          new SelectCustomerListHolder(
              this._fragment, CustomerCardWideBinding.inflate(inflater, parent, false));
    };
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerViewHolder holder, int index) {
    Objects.requireNonNull(holder);

    if (holder instanceof SelectCustomerHeaderHolder headerHolder) {
      final Optional<CustomerModel> selectedCustomer =
          Optional.ofNullable(this._fragment.selectCustomerViewModel().initialSelectedCustomer());
      headerHolder.bind(selectedCustomer);

    } else if (holder instanceof SelectCustomerListHolder listHolder) {
      // -1 offset because header holder.
      listHolder.bind(
          this._fragment.selectCustomerViewModel().customers().getValue().get(index - 1));
    }
  }

  @Override
  public int getItemCount() {
    // +1 offset because header holder.
    return this._fragment.selectCustomerViewModel().customers().getValue().size() + 1;
  }

  @Override
  public int getItemViewType(int index) {
    return switch (index) {
      case 0 -> ViewType.HEADER.value();
      default -> ViewType.LIST.value();
    };
  }
}

File: app/src/main/java/com/robifr/ledger/ui/selectcustomer/recycler/SelectCustomerAdapter.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.selectcustomer.recycler

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.robifr.ledger.databinding.CustomerCardWideBinding
import com.robifr.ledger.databinding.ListableListSelectedItemBinding
import com.robifr.ledger.ui.RecyclerViewHolderKt
import com.robifr.ledger.ui.selectcustomer.SelectCustomerFragment

class SelectCustomerAdapter(private val _fragment: SelectCustomerFragment) :
    RecyclerView.Adapter<RecyclerViewHolderKt>() {
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolderKt {
    return when (ViewType.entries.find { it.value == viewType }) {
      ViewType.HEADER ->
          SelectCustomerHeaderHolder(
              _selectedItemBinding =
                  ListableListSelectedItemBinding.inflate(_fragment.layoutInflater, parent, false),
              _initialSelectedCustomer = {
                _fragment.selectCustomerViewModel.uiState.safeValue.initialSelectedCustomer
              },
              _isSelectedCustomerPreviewExpanded = {
                _fragment.selectCustomerViewModel.uiState.safeValue
                    .isSelectedCustomerPreviewExpanded
              },
              _onSelectedCustomerPreviewExpanded =
                  _fragment.selectCustomerViewModel::onSelectedCustomerPreviewExpanded)
      else ->
          SelectCustomerListHolder(
              _cardBinding =
                  CustomerCardWideBinding.inflate(_fragment.layoutInflater, parent, false),
              _initialSelectedCustomerIds = {
                listOfNotNull(
                    _fragment.selectCustomerViewModel.uiState.safeValue.initialSelectedCustomer?.id)
              },
              _customers = { _fragment.selectCustomerViewModel.uiState.safeValue.customers },
              _onCustomerSelected = _fragment.selectCustomerViewModel::onCustomerSelected,
              _expandedCustomerIndex = {
                _fragment.selectCustomerViewModel.uiState.safeValue.expandedCustomerIndex
              },
              _onExpandedCustomerIndexChanged =
                  _fragment.selectCustomerViewModel::onExpandedCustomerIndexChanged)
    }
  }

  override fun onBindViewHolder(holder: RecyclerViewHolderKt, position: Int) {
    when (holder) {
      is SelectCustomerHeaderHolder -> holder.bind()
      is SelectCustomerListHolder -> holder.bind(position - 1)
    }
  }

  override fun getItemCount(): Int =
      // +1 offset because header holder.
      _fragment.selectCustomerViewModel.uiState.safeValue.customers.size + 1

  override fun getItemViewType(position: Int): Int =
      when (position) {
        0 -> ViewType.HEADER.value
        else -> ViewType.LIST.value
      }

  private enum class ViewType(val value: Int) {
    HEADER(0),
    LIST(1)
  }
}
